### PR TITLE
Compressing tree-sequences to stdout

### PR DESCRIPTION
**Pull request by aabiddanda**
In response to #49, I've implemented a version (& some testing) that allows one to compress a single tree to `stdout` and pipe to a file of choice. The code is a bit clunky w.r.t. filehandling so I'm making this as a draft PR initially to get some feedback. 

There is one outstanding thing that I have not added, and that is how to compress multiple tree sequences to the stdout stream and verify that we can read them in sequence (similar to sequential calls to `tskit.load` from a single file).

### CODE REVIEW

**Comment by jeromekelleher** [comment]
Thanks for following up on this @aabiddanda!

Have you tried passing ``sys.stdout`` to zarr directly? It looks like me like the zipfile interface [https://docs.python.org/3.7/library/zipfile.html#zipfile.ZipFile] supports this, and the [zarr code](https://github.com/zarr-developers/zarr-python/blob/a92550b9ff9f4931353b4066486aaf9755895ce3/zarr/storage.py#L1513) just passes the ``path`` argument through. So, ideally, we could pass the file handle through directly?

**Comment by aabiddanda** [comment]
I've tried every permutation of trying to get `sys.stdout` and `sys.stdout.buffer` passed directly as a `zipfile` (which the naive `ZipFile` interface allows), but I think that the broader issue with zarr is that it expects a full string path (see [line 1500 here](https://github.com/zarr-developers/zarr-python/blob/a92550b9ff9f4931353b4066486aaf9755895ce3/zarr/storage.py#L1500-L1513)).  This isn't something that can easily be solved by just passing in `sys.stdout` or even `/dev/stdout` (which leads to hanging on the file handle on my Mac). Would appreciate any suggestions @jeromekelleher if you have insights here.

File: tszip/compression.py [hunk 1]
```diff
@@ -95,8 +97,13 @@ def compress(ts, path, variants_only=False):
         with zarr.ZipStore(filename, mode="w") as store:
             root = zarr.group(store=store)
             compress_zarr(ts, root, variants_only=variants_only)
-        os.replace(filename, destination)
-    logging.info(f"Wrote {destination}")
+        if stdout:
```
Comment (jeromekelleher) [inline tszip/compression.py:100]: Ah, looks like we're actually writing a temporary file anyway, so it's not that bad. How about this (untested!)

```python
if isinstance(destination, (str, pathlib.Path)):
     os.replace(filename, destination)
     logging.info(f"Wrote {destination}")
else:
     # Assume that destination is a file-like object open in "wb" mode.
     with open(filename, "rb") as source:
            chunk_size = 2**10  # 1MiB
            for chunk in iter(functools.partial(source.read, 64), b""):
                  destination.write(chunk)
```

That way we don't need the stdout option and can write to any suitable file-like object.

**Review by jeromekelleher — COMMENTED** [review]
What do you think of this version @aabiddanda?

File: tszip/compression.py (hunk 1 above)
Comment (aabiddanda) [reply tszip/compression.py:100]: In general I like this solution as it helps keep the function signature the same. I think it takes a little bit of messing with because it does not reliably work when constructing an in-place directory when the path being fed in is `sys.stdout`, but I think that it could be worked around. I'll see what I can do for this to make it work with the various tests in place as well. 
```
    destination = str(path)
    # Write the file into a temporary directory on the same file system so that
    # we can write the output atomically.
    destdir = os.path.dirname(os.path.abspath(destination))
    with tempfile.TemporaryDirectory(dir=destdir, prefix=".tszip_work_") as tmpdir:
```

File: tszip/compression.py [hunk 2]
```diff
@@ -85,25 +86,26 @@ def compress(ts, path, variants_only=False, stdout=False):
...
-    :param bool stdout: If True, write the tree sequence to stdout.
     """
-    destination = str(path)
+    destination = path
     # Write the file into a temporary directory on the same file system so that
     # we can write the output atomically.
-    destdir = os.path.dirname(os.path.abspath(destination))
-    with tempfile.TemporaryDirectory(dir=destdir, prefix=".tszip_work_") as tmpdir:
+    # destdir = os.path.dirname(os.path.abspath(destination))
+    with tempfile.TemporaryDirectory(dir=None, prefix=".tszip_work_") as tmpdir:
```
Comment (jeromekelleher) [inline tszip/compression.py:94]: This is almost right I think, but it does break the atomic write contract which ensures that we don't leave partially written files lying around. The idea of putting the temporary file in the same directory as the destination is that moving the temporary file to it's final destination is simply a rename, which is as close to atomic as you can get. However, if the temporary file is in the standard temp directory, then this becomes a real move because ``/tmp`` is usually a in a different file system.

So, how about this:

```python
try:
    destination = pathlib.Path(path).resolve()
    is_path = True
    tmpdir = destination.parent # I think this is right?
except TypeError:
    is_path = False 
    tmpdir = None
with tempfile.TemporaryDirectory(dir=tmpdir, prefix=".tszip_work_") as tmpdir:
      # same as before 
      filename = pathlib.Path(tmpdir) / "tmp.trees.tgz"
      with zarr.ZipStore(filename, mode="w") as store:
          # blah blah
      if is_path:
          os.replace(..) # atomic file move
      else:
          # write chunks to dest
```
```

Comment (jeromekelleher) [reply tszip/compression.py:94]: Seems like a good idea to update to pathlib here as well while we're at it? That's no necessary, though.

Comment (aabiddanda) [reply tszip/compression.py]: Great suggestion! This solution works really great and actually passes all of the tests that we've put forth.

There is a somewhat insidious bug when actually running on the command line: 
```
pip install .
tszip -c tests/files/1.0.0.trees
```
though because `sys.stdout` expects strings (but in the tests we overload `sys.stdout` to be a buffer in `wb+` mode) for capturing output in our pytest suite...

Comment (jeromekelleher) [reply tszip/compression.py]: That's weird - what happens when we write to stdout? It seems to work OK when we write binary data to stdout in msprime (try ``msp ancestry``) and it doesn't look like there's any changes made there... Ah, maybe because we're writing directly to the file descriptor there and not doing it in Python.

OK, trying ``sys.stdout.buffer`` as the argument?

**Comment by jeromekelleher** [comment]
Thanks for picking this up again @aabiddanda! I've added a few small tweaks, what do you think? Basically I don't like the idea of complicating the library code just to make the tests work - I think this is cleaner, even if the tests are a bit more complicated.

If you're happy, then can you rebase and squash the commits down to one so we can merge please?

**Comment by aabiddanda** [comment]
Should be squashed and good to go.  

However, I think that it should be noted that the behavior for compressing multiple files to stdout is a little bit different from `gzip` when using the `-c` flag. When testing this, it seems that only the final compressed tree is available. For example, a test like this fails: 
```
    def test_compress_stdout_multiple(self):
        self.assertTrue(self.trees_path.exists())
        tmp_file = pathlib.Path(self.tmpdir.name) / "stdout_mult.trees.tsz"
        with mock.patch("tszip.cli.get_stdout", wraps=get_stdout_for_pytest):
            stdout, stderr = self.run_tszip_stdout(["-c", str(self.trees_path), str(self.trees_path2)])
        with open(tmp_file, "wb+") as tmp:
            tmp.write(stdout)
        ts_mult1 = tszip.decompress(tmp_file)
        ts_mult2 = tszip.decompress(tmp_file)
        # We only get the last one?
        self.assertEqual(ts_mult1.tables, self.ts.tables) # this line fails
        self.assertEqual(ts_mult2.tables, self.ts2.tables)
```

If we only can compress one tree-sequence file at a time on the commandline, then we should potentially raise a warning if a user tries to compress multiple tree-sequences to a single file?

**Comment by jeromekelleher** [comment]
Looks like we still have 14 commits @aabiddanda, so maybe you didn't force push or something? This [guide](https://popsim-consortium.github.io/stdpopsim-docs/stable/development.html#rebasing) should help, but I'm happy to do the squashing if you prefer.

Re the multiple files thing - hmm, good catch. Maybe we should merge this much first, and open a separate issue to track that problem? We can then decide how to deal with it before releasing. (Just raising an error is fine if we need to IMO - it's an edge case)

**Comment by aabiddanda** [comment]
Alright I think that its now squashed to one commit @jeromekelleher (I'm always amazed by the rebasing process). 

Re: multiple files I agree with your suggestion of making a separate issue and adding in a test potentially for that edge case.

**Review by jeromekelleher — APPROVED** [review]
This is great, for taking the time to work this through @aabiddanda!
